File: scripts/dsa_lista_modelos.mjs
// Módulo Especial de Consultoria na Área de Dados com Agentes de IA
// Utilitario de apoio: lista os modelos VALIDOS da sua conta em cada provedor.
//
// VERSAO NODE.JS -- corresponde ao dsa_lista_modelos.py.
//
// Para que serve: nomes de modelo saem do ar sem aviso. Foi o que aconteceu com
// os dois modelos originais deste capitulo -- "deepseek-r1-distill-llama-70b" e
// "llama-3.3-70b-versatile" hoje retornam erro na Groq. Quando o app reclamar
// que o modelo nao existe, rode este script, escolha um nome da lista e coloque
// no .env (ou nas variaveis de ambiente da Vercel).
//
// Uso:  npm run modelos
//       (que executa: node --env-file-if-exists=.env scripts/dsa_lista_modelos.mjs)
//
// O que mudou em relacao ao Python: la usavamos o SDK de cada provedor
// (groq, openai, anthropic, ollama). Aqui usamos so o 'fetch' nativo do Node,
// porque a rota "listar modelos" e um GET simples em todos eles -- e assim
// voce ve o endereco real de cada API. O .env e carregado pelo proprio Node
// (--env-file), sem dotenv.
//
// A extensao .mjs diz ao Node que o arquivo usa 'import' (modulos ES) em vez
// do 'require' antigo.

// Mesma regra do app: chave vazia ou com texto de exemplo = nao configurada.
const DSA_PLACEHOLDERS = ["seu-", "sua-", "cole-", "coloque", "your-", "xxx", "<"];

function dsaChaveValida(nomeDaVariavel) {
  const valor = (process.env[nomeDaVariavel] ?? "").trim().replace(/^["']|["']$/g, "");
  if (!valor || DSA_PLACEHOLDERS.some((inicio) => valor.toLowerCase().startsWith(inicio))) return null;
  return valor;
}

function titulo(texto) {
  console.log();
  console.log("=".repeat(78));
  console.log(texto);
  console.log("=".repeat(78));
}

/** GET com cabecalhos; lanca erro com o corpo da resposta se nao for 2xx. */
async function dsaGetJson(url, cabecalhos = {}) {
  const resposta = await fetch(url, { headers: cabecalhos, signal: AbortSignal.timeout(15_000) });
  if (!resposta.ok) throw new Error(`HTTP ${resposta.status}: ${(await resposta.text()).slice(0, 200)}`);
  return resposta.json();
}

async function listaGroq() {
  titulo("GROQ");
  const chave = dsaChaveValida("GROQ_API_KEY");
  if (!chave) return console.log("  GROQ_API_KEY nao configurada no .env -- provedor pulado.");
  const { data } = await dsaGetJson("https://api.groq.com/openai/v1/models", { Authorization: `Bearer ${chave}` });
  const modelos = data.map((m) => m.id).sort();
  for (const m of modelos) console.log("  -", m);
  console.log(`  (${modelos.length} modelos)`);
}

async function listaOpenAI() {
  titulo("OPENAI");
  const chave = dsaChaveValida("OPENAI_API_KEY");
  if (!chave) return console.log("  OPENAI_API_KEY nao configurada no .env -- provedor pulado.");
  const { data } = await dsaGetJson("https://api.openai.com/v1/models", { Authorization: `Bearer ${chave}` });
  const modelos = data.map((m) => m.id).sort();
  // A conta costuma ter dezenas de modelos (audio, imagem, embeddings...).
  // Mostramos so os de conversa, que sao os que interessam para os agentes.
  const conversa = modelos.filter((m) => /^(gpt-|o1|o3|o4)/.test(m));
  for (const m of conversa) console.log("  -", m);
  console.log(`  (${conversa.length} de conversa, ${modelos.length} no total)`);
}

async function listaAnthropic() {
  titulo("ANTHROPIC");
  const chave = dsaChaveValida("ANTHROPIC_API_KEY");
  if (!chave) return console.log("  ANTHROPIC_API_KEY nao configurada no .env -- provedor pulado.");
  // A Anthropic usa o cabecalho x-api-key (nao Authorization) e exige a versao da API.
  const { data } = await dsaGetJson("https://api.anthropic.com/v1/models?limit=100", {
    "x-api-key": chave,
    "anthropic-version": "2023-06-01",
  });
  const modelos = data.map((m) => m.id);
  for (const m of modelos) console.log("  -", m);
  console.log(`  (${modelos.length} modelos)`);
}

async function listaGrok() {
  titulo("GROK (xAI)");
  const chave = dsaChaveValida("XAI_API_KEY");
  if (!chave) return console.log("  XAI_API_KEY nao configurada no .env -- provedor pulado.");
  // A xAI fala o MESMO protocolo da OpenAI. So muda o endereco do servidor.
  const { data } = await dsaGetJson("https://api.x.ai/v1/models", { Authorization: `Bearer ${chave}` });
  const modelos = data.map((m) => m.id).sort();
  for (const m of modelos) console.log("  -", m);
  console.log(`  (${modelos.length} modelos)`);
}

async function listaOllama() {
  titulo("OLLAMA (local, sem chave)");
  const endereco = process.env.OLLAMA_HOST || "http://localhost:11434";
  let modelos;
  try {
    ({ models: modelos } = await dsaGetJson(`${endereco}/api/tags`));
  } catch (erro) {
    console.log(`  Servidor local nao respondeu em ${endereco}: ${erro.message}`);
    console.log("  Dica: instale com 'curl -fsSL https://ollama.com/install.sh | sh'");
    return;
  }
  for (const m of modelos) console.log("  -", m.model);
  console.log(`  (${modelos.length} modelos baixados em ${endereco})`);
}

for (const funcao of [listaGroq, listaOpenAI, listaAnthropic, listaGrok, listaOllama]) {
  try {
    await funcao();
  } catch (erro) {
    console.log(`  ERRO ao consultar: ${erro.constructor.name}: ${erro.message}`);
  }
}
console.log();
console.log("Copie o nome escolhido para a variavel de modelo no .env.");
console.log("Ex.: GROQ_MODEL=openai/gpt-oss-120b");

File: dsa_lista_modelos.py
# Módulo Especial de Consultoria na Área de Dados com Agentes de IA
# Utilitario de apoio: lista os modelos VALIDOS da sua conta em cada provedor.
#
# Para que serve: nomes de modelo saem do ar sem aviso. Foi o que aconteceu com
# os dois modelos originais deste capitulo -- "deepseek-r1-distill-llama-70b" e
# "llama-3.3-70b-versatile" hoje retornam erro 404 na Groq. Quando o app
# reclamar de "model_not_found", rode este script, escolha um nome da lista e
# coloque no .env.
#
# Uso:  uv run python dsa_lista_modelos.py

import os
from pathlib import Path

from dotenv import load_dotenv

load_dotenv(Path(__file__).parent / ".env")

# Mesma regra do app: chave vazia ou com texto de exemplo = nao configurada.
DSA_PLACEHOLDERS = ("seu-", "sua-", "cole-", "coloque", "your-", "xxx", "<")


def dsa_chave_valida(nome_da_variavel):
    valor = (os.getenv(nome_da_variavel) or "").strip().strip('"').strip("'")
    if not valor or valor.lower().startswith(DSA_PLACEHOLDERS):
        return None
    return valor


def titulo(texto):
    print()
    print("=" * 78)
    print(texto)
    print("=" * 78)


def lista_groq():
    titulo("GROQ")
    chave = dsa_chave_valida("GROQ_API_KEY")
    if not chave:
        print("  GROQ_API_KEY nao configurada no .env -- provedor pulado.")
        return
    from groq import Groq

    modelos = sorted(m.id for m in Groq(api_key=chave).models.list().data)
    for m in modelos:
        print("  -", m)
    print(f"  ({len(modelos)} modelos)")


def lista_openai():
    titulo("OPENAI")
    chave = dsa_chave_valida("OPENAI_API_KEY")
    if not chave:
        print("  OPENAI_API_KEY nao configurada no .env -- provedor pulado.")
        return
    from openai import OpenAI

    modelos = sorted(m.id for m in OpenAI(api_key=chave).models.list().data)
    # A conta costuma ter dezenas de modelos (audio, imagem, embeddings...).
    # Mostramos so os de conversa, que sao os que interessam para os agentes.
    conversa = [m for m in modelos if m.startswith(("gpt-", "o1", "o3", "o4"))]
    for m in conversa:
        print("  -", m)
    print(f"  ({len(conversa)} de conversa, {len(modelos)} no total)")


def lista_anthropic():
    titulo("ANTHROPIC")
    chave = dsa_chave_valida("ANTHROPIC_API_KEY")
    if not chave:
        print("  ANTHROPIC_API_KEY nao configurada no .env -- provedor pulado.")
        return
    import anthropic

    modelos = [m.id for m in anthropic.Anthropic(api_key=chave).models.list().data]
    for m in modelos:
        print("  -", m)
    print(f"  ({len(modelos)} modelos)")


def lista_grok():
    titulo("GROK (xAI)")
    chave = dsa_chave_valida("XAI_API_KEY")
    if not chave:
        print("  XAI_API_KEY nao configurada no .env -- provedor pulado.")
        return
    # A xAI usa a MESMA biblioteca da OpenAI. So muda o endereco do servidor.
    from openai import OpenAI

    cliente = OpenAI(api_key=chave, base_url="https://api.x.ai/v1")
    modelos = sorted(m.id for m in cliente.models.list().data)
    for m in modelos:
        print("  -", m)
    print(f"  ({len(modelos)} modelos)")


def lista_ollama():
    titulo("OLLAMA (local, sem chave)")
    import ollama

    endereco = os.getenv("OLLAMA_HOST") or "http://localhost:11434"
    try:
        modelos = ollama.Client(host=endereco, timeout=10).list()["models"]
    except Exception as erro:
        print(f"  Servidor local nao respondeu em {endereco}: {erro}")
        print("  Dica: instale com 'curl -fsSL https://ollama.com/install.sh | sh'")
        return
    for m in modelos:
        print("  -", m["model"])
    print(f"  ({len(modelos)} modelos baixados em {endereco})")


if __name__ == "__main__":
    for funcao in (lista_groq, lista_openai, lista_anthropic, lista_grok, lista_ollama):
        try:
            funcao()
        except Exception as erro:
            print(f"  ERRO ao consultar: {type(erro).__name__}: {erro}")
    print()
    print("Copie o nome escolhido para a variavel de modelo no .env.")
    print("Ex.: GROQ_MODEL=openai/gpt-oss-120b")
